### PR TITLE
[TE] Implementa a workaround solution for no-auth on new create function

### PR DESCRIPTION
**Pull request by cyy0714**
During onboard, we found that new created functions are without owner information. This is because of the credential information cannot pass to workers. A temporary solution is posted to create a dummy function on dashboard, and run the onboarding pipelines on the dummy function. Some fixes are also pushed to resolve null pointer exceptions under corner cases.

### CODE REVIEW

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/alert/content/EmailContentFormatterContext.java [hunk 1]
```diff
@@ -1,12 +1,32 @@
 package com.linkedin.thirdeye.alert.content;
 
+import com.linkedin.thirdeye.datalayer.dto.AlertConfigDTO;
+import com.linkedin.thirdeye.datalayer.dto.AnomalyFunctionDTO;
 import org.joda.time.DateTime;
 
 
 public class EmailContentFormatterContext {
+  private AnomalyFunctionDTO anomalyFunctionSpec;
+  private AlertConfigDTO alertConfig;
```
Comment (apucher) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/alert/content/EmailContentFormatterContext.java]: How is this related to the dummy-function workaround?

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/alert/content/OnboardingNotificationEmailContentFormatter.java [hunk 2]
```diff
@@ -71,12 +71,9 @@ protected void updateTemplateDataByAnomalyResults(Map<String, Object> templateDa
...
     }
-    AlertConfigDTO alertConfig = null;
     templateData.put("alertPattern", alertPattern);
-    if (templateData.containsKey(ALERT_CONFIG_NAME)) {
-      String configName = (String) templateData.get(ALERT_CONFIG_NAME);
-      alertConfig = DAORegistry.getInstance().getAlertConfigDAO().findWhereNameEquals(configName);
-    }
+
+    AlertConfigDTO alertConfig = context.getAlertConfig();
     if (alertConfig == null) {
```
Comment (apucher) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/alert/content/OnboardingNotificationEmailContentFormatter.java]: same

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/onboard/FunctionOnboardingResource.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+import com.linkedin.thirdeye.datasource.DAORegistry;
+import javax.validation.constraints.NotNull;
+import javax.ws.rs.POST;
+import javax.ws.rs.Path;
+import javax.ws.rs.Produces;
+import javax.ws.rs.QueryParam;
+import javax.ws.rs.core.MediaType;
+
+
+@Path("/detection-onboard")
```
Comment (apucher) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/onboard/FunctionOnboardingResource.java]: this namespace clashes with the existing "detection-onboard" resource and/or the reverse proxy. Please choose a different name.

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/onboard/tasks/FunctionCreationOnboardingTask.java [hunk 4]
```diff
@@ -179,27 +184,45 @@ public void run() {
         defaultFunctionProperties.setProperty((String) propertyEntry.getKey(), (String) propertyEntry.getValue());
       }
 
-      Response response = anomalyResource.createAnomalyFunction(configuration.getString(COLLECTION_NAME),
```
Comment (apucher) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/onboard/tasks/FunctionCreationOnboardingTask.java]: thanks for cleaning this up

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/onboard/tasks/FunctionCreationOnboardingTask.java [hunk 5]
```diff
@@ -179,27 +184,45 @@ public void run() {
...
+      anomalyFunction.setBucketSize(dataGranularity.getSize());
+      anomalyFunction.setBucketUnit(dataGranularity.getUnit());
+      anomalyFunction.setIsActive(configuration.getBoolean(IS_ACTIVE, DEFAULT_IS_ACTIVE));
+
+      anoomalyFunctionDAO.update(anomalyFunction);
+
+      executionContext.setExecutionResult(ANOMALY_FUNCTION_CONFIG, anomalyFunction);
     } catch (Exception e) {
-      throw new IllegalArgumentException(e);
+      throw new IllegalArgumentException(e.fillInStackTrace());
```
Comment (apucher) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/onboard/tasks/FunctionCreationOnboardingTask.java]: the original, nested exception passing, is the correct way of doing this

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/onboard/tasks/FunctionReplayOnboardingTask.java [hunk 6]
```diff
@@ -65,7 +65,7 @@ public void run() {
       }
 
     } catch (Exception e) {
-      throw new IllegalStateException(e);
+      throw new IllegalStateException(e.fillInStackTrace());
```
Comment (apucher) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/onboard/tasks/FunctionReplayOnboardingTask.java]: same

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/alert/content/EmailContentFormatterContext.java (hunk 1 above)
Comment (cyy0714) [reply thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/alert/content/EmailContentFormatterContext.java]: This is not related to the workaround. It is to fix when there is no anomaly generated, we might face null pointer exception when generating emails.

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/onboard/FunctionOnboardingResource.java (hunk 3 above)
Comment (cyy0714) [reply thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/onboard/FunctionOnboardingResource.java]: Change to function-onboard.

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/onboard/tasks/FunctionCreationOnboardingTask.java (hunk 5 above)
Comment (cyy0714) [reply thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/onboard/tasks/FunctionCreationOnboardingTask.java]: Thanks for the comment. It's been fixed.
